src/odemis/driver/tmcm.py

```diff
@@ -2502,3 +2553,871 @@ def _parseMessage(self, msg):
+        object (0x607C) to the current position after the homing procedure (default: true)
+        :return bool: If the homing was complete with success
+        """
+        # Use init lock to make sure we're not cancelling during the state changes.
+        with future._init_lock:
+            if future._must_stop.is_set():
+                # Cancelled before homing started (init_lock acquired here after cancelling function)
+                return False
+            previus_op_mode = self._node.op_mode
+            self._node.state = 'SWITCHED ON'
+            self._node.op_mode = 'HOMING'
+            # The homing process will initialize at operation enabled
+            self._node.state = 'OPERATION ENABLED'
+            homingstatus = 'IN PROGRESS'
+            self._node.controlword = State402.CW_OPERATION_ENABLED | Homing.CW_START
+        t = time.time() + timeout
+        try:
+            while homingstatus not in ('TARGET REACHED', 'ATTAINED'):
+                for key, value in Homing.STATES.items():
+                    # check if the value after applying the bitmask (value[0])
+                    # corresponds with the value[1] to determine the current status
+                    bitmaskvalue = self._node.statusword & value[0]
+                    if bitmaskvalue == value[1]:
+                        homingstatus = key
+                if homingstatus in ('INTERRUPTED', 'ERROR VELOCITY IS NOT ZERO', 'ERROR VELOCITY IS ZERO'):
+                    raise RuntimeError('Unable to home. Reason: {0}'.format(homingstatus))
+                time.sleep(0.001)
+                if time.time() > t:
+                    raise RuntimeError('Unable to home, timeout reached')
+            if set_new_home:
+                actual_position = self._node.sdo[ACTUAL_POSITION].raw
+                self._node.sdo[HOME_OFFSET].raw = actual_position  # home offset (0x607C)
+                logging.info('Homing offset set to {0}'.format(actual_position))
+            logging.info('Homing mode carried out successfully.')
+            return True
+        except RuntimeError as e:
+            logging.info(str(e))
+        finally:
+            self._node.op_mode = previus_op_mode
+        return False
+
+    @staticmethod
+    def _openCanNode(channel, nodeid, datasheet):
+        """
+        Create a single-node network.
+
+        raise HwError: if the serial port cannot be opened (doesn't exist, or
+          already opened)
+        """
+        # For debugging purpose
+        if channel == "fake":
+            return None, CANNodeSimulator(naxes=1)
+
+        # Start with creating a network representing one CAN bus
+        network = canopen.Network()
+
+        # Connect to the CAN bus
+        network.connect(bustype='socketcan', channel=channel)
+        network.check()
+
+        # Add some nodes with corresponding Object Dictionaries
+        node = canopen.BaseNode402(nodeid, datasheet)
+        network.add_node(node)
```

Comment:
Like this?
```
        # Connect to the CAN bus
        try:
            network.connect(bustype='socketcan', channel=channel)
            network.check()
        except CanError:
            raise HwError("Can adapter not found.")

        # Add some nodes with corresponding Object Dictionaries
        node = canopen.BaseNode402(nodeid, datasheet)
        network.add_node(node)

        # Reset network
        node.nmt.state = 'RESET COMMUNICATION'
        node.nmt.wait_for_bootup(15)
        logging.debug('Device state after reset = {0}'.format(node.nmt.state))

        # Transmit SYNC every 100 ms
        network.sync.start(0.1)

        try:
            node.load_configuration()
            node.setup_402_state_machine()
        except ValueError:
            raise HwError("Node not present on the bus.")
```

Reply:
Yes... I trust you've tested it with both cases and that's indeed the exceptions that happen in both cases :-)

Just a small request, please add a bit more info in the error messages. Indicate the "channel" in the CAN adapter error (and also spell "CAN" in uppercase, to avoid confusion). For the node missing, please add both the channel and nodeid.